src/contracts/fungible/cache/file.rs

```diff
@@ -192,4 +192,178 @@ impl Cache for FileCache {
         self.save()?;
         Ok(existed)
     }
+
+    fn utxo_allocation_map(
+        &self,
+        contract_id: ContractId,
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, CacheError> {
+        let asset = self.asset(contract_id).unwrap();
+
+        let allocation_map = asset.known_allocations();
+
+        let mut result = HashMap::new();
+
+        for item in allocation_map {
+            let mut alloc_amount = vec![];
+            for alloc in item.1 {
+                alloc_amount.push(alloc.value().value);
+            }
+            result.insert(item.0.clone(), alloc_amount);
+        }
+
+        Ok(result)
+    }
+
+    fn asset_allocation_map(
+        &self,
+        utxo: &bitcoin::OutPoint,
+    ) -> Result<HashMap<ContractId, Vec<AtomicValue>>, CacheError> {
+        let mut result = HashMap::new();
+
+        for asset in self
+            .assets()
+            .unwrap()
+            .into_iter()
+            .cloned()
+            .collect::<Vec<Asset>>()
+        {
+            let allocations: Vec<AtomicValue> =
+                match asset.known_allocations().get(utxo) {
+                    Some(allocations) => allocations
+                        .into_iter()
+                        .map(|alloc| alloc.value().value)
+                        .collect(),
+                    None => continue,
+                };
+
+            result.insert(asset.id().clone(), allocations);
+        }
```

Comment:
This also can be done without any single clone (btw ContractId is a copy type)

Reply:
Ya its working without `clone()` but i need to at least deref the value because `asset.id()` gives a `&ContractId`